cu29/src/cutask.rs

```diff
@@ -16,8 +16,51 @@ use std::fmt::{Display, Formatter};
 // Everything that is stateful in copper for zero copy constraints need to be restricted to this trait.
 pub trait CuMsgPayload: Default + Encode + Decode + Sized {}
 
+pub trait CuMsgPack<'cl> {}
+
 // Also anything that follows this contract can be a payload (blanket implementation)
-impl<T> CuMsgPayload for T where T: Default + Encode + Decode + Sized {}
+impl<T: Default + Encode + Decode + Sized> CuMsgPayload for T {}
+
+macro_rules! impl_cu_msg_pack {
+    ($(($($ty:ident),*)),*) => {
+        $(
+            impl<'cl, $($ty: CuMsgPayload + 'cl),*> CuMsgPack<'cl> for ( $( &'cl CuMsg<$ty>, )* ) {}
+        )*
+    };
+}
+
+impl<'cl, T: CuMsgPayload> CuMsgPack<'cl> for (&'cl CuMsg<T>,) {}
+impl<'cl, T: CuMsgPayload> CuMsgPack<'cl> for &'cl CuMsg<T> {}
+impl<'cl, T: CuMsgPayload> CuMsgPack<'cl> for (&'cl mut CuMsg<T>,) {}
+impl<'cl, T: CuMsgPayload> CuMsgPack<'cl> for &'cl mut CuMsg<T> {}
+impl<'cl> CuMsgPack<'cl> for () {}
+
+// Apply the macro to generate implementations for tuple sizes up to 5
+impl_cu_msg_pack! {
+    (T1, T2), (T1, T2, T3), (T1, T2, T3, T4), (T1, T2, T3, T4, T5) // TODO: continue if necessary
```

Comment:
Can we automate this with something like https://docs.rs/variadics/latest/variadics/#var_type?

Reply:
I looked into it a little bit but those recursive constructs would be exposed to the users I believe: instead of a tuple of &CuMsg you have a (&CuMsg, (&CuMsg, )) so you need to also have the user use the macros from variadics to unpack them.

I would say, let's wait that it becomes standard in Rust so we don't paint ourselves in a corner.